playbooks/upload_to_swift.yml

```diff
@@ -62,23 +101,107 @@
-      template:
-        src: swift.html.j2
-        dest: "{{ report_dir }}/index.html"
+
+    # In order for uploaded files to be browsable:
+    # 1. Static hosting must be enabled (this links index.html to requests for container/)
+    # 2. CDN must be enabled (to allow anonymous http access to the container)
+    # 3. An index page must be generated, as this is not done automatically
+    - name: Enable static web hosting
+      uri:
+        url: "{{object_store_url}}"
+        method: POST
+        headers:
+          X-AUTH-TOKEN: "{{ auth_token }}"
+          X-Container-Meta-Web-Index: index.html
+        status_code: 200, 201, 202, 204
+      no_log: true
+
+    # Do this before generating index.html, styles.css and data.json
+    # So they don't show up on the final page.
+    - name: Generate file list with sizes for index page data
+      shell: |
+        find . \
+          -type f \
+          -mindepth 1 \
+          -exec ls -l {} \; \
+        |sed 's+\./++g'
+      args:
+        chdir: "{{ adir }}"
+      register: file_list
+
+    - name: Copy index.html
+      copy:
+        src: templates/artifact/index.html
+        dest: "{{ adir }}/index.html"
+
+    - name: Copy style.css
+      copy:
+        src: templates/artifact/styles.css
+        dest: "{{ adir }}/styles.css"
+
+    # this is json data containing metadata and a list of files
+    # that will be downloaded by index.html
+    # and used to display a list of available artifacts
+    - name: Generate data.json
+      copy:
+        content: |
+          {
+            "job_name": "{{ job_name }}",
+            "build_number": "{{ build_number }}",
+            "archive_base_name": "{{ archive_base_name }}",
+            "container_public_url": "{{ container_public_url}}",
+            "files": [
+              {% for file_line in file_list.stdout_lines %}
+                {% set file_list = file_line.split() %}
+                {% if file_list|length > 3 %}
+                  {
+                    "path": "{{file_list[-1]}}",
+                    "size": {{file_list[4]}}
+                  }{{ "," if not loop.last else "" }}
+                {% endif %}
+              {% endfor %}
+            ],
+            "archives": [
+              "{{ container_public_url }}/{{ archive_base_name }}.tar.gz"
+            ]
+          }
+        dest: "{{ adir }}/data.json"
+
+    # os_object does not currently support setting object expiration header field
+    # production retention
+    - name: Upload Artifacts to Cloud Files
+      command: "swift upload --object-threads 100 --ignore-checksum --header 'X-Delete-After:{{ retention }}' {{ container }} {{ adir | basename }}"
+      args:
+        chdir: "{{ adir |dirname }}"
+      environment:
+        OS_AUTH_TOKEN: "{{ auth_token }}"
+        OS_STORAGE_URL: "{{ (object_store['endpoints'] | selectattr('region', 'equalto', region) | first)['publicURL'] }}"
+      failed_when: false
```

Comment:
Why are we forcing this never to fail? If there's a legit reason, we should probably add a comment as to why.

Does the swift client do retries on its own, or shouldn't this task do retries if there's a failure? If the swift client retries, does it see whether some files are there already and skip them?

Reply:
Ah ok, so sift continues uploading anything else. Sounds good then. Is there a way to capture the failed files (if there are any) so that we can also put that into a file to show?